File: src/main/kotlin/io/titandata/titan/commands/Migrate.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.commands

import io.titandata.titan.Dependencies
import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.requireObject
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.options.required
import org.kodein.di.Kodein
import org.kodein.di.generic.bind
import org.kodein.di.generic.inSet
import org.kodein.di.generic.provider

private val n = System.lineSeparator()

class Migrate : CliktCommand(
        help = "Migrate an existing docker database container to titan repository",
        epilog = "Container becomes the new name of the docker container.${n}${n}Example: `titan migrate -s oldPostgres titanPostgres`"
) {
    private val dependencies: Dependencies by requireObject()
    private val repository by argument()
    private val source by option("-s", "--source", help="Required. Source docker database container").required()
    override fun run() {
        val provider = dependencies.provider
        provider.migrate(source, repository)
    }
}

val migrateModule = Kodein.Module("migrate") {
    bind<CliktCommand>().inSet() with provider {
        Migrate()
    }
}

File: internal/app/commands/migrate.go
/*
Copyright © 2019 The Titan Project Contributors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package commands

import (
	"github.com/spf13/cobra"
)

// migrateCmd represents the migrate command
var migrateCmd = &cobra.Command{
	Use:   "migrate [REPOSITORY]",
	Short: "Migrate an existing docker database container to titan repository",
	Long: `Migrate an existing docker database container to titan repository. 
Container becomes the new name of the docker container.

Example: 'titan migrate -s oldPostgres titanPostgres'`,
	Args: cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		repo := args[0]
		provider.Migrate(source, repo)
	},
}

func init() {
	rootCmd.AddCommand(migrateCmd)
	migrateCmd.Flags().StringVarP(&source, "source", "s", "", "source docker database container (required)")
}
